stl/inc/format

```diff
@@ -38,36 +40,248 @@ enum class _Align { _None, _Left, _Right, _Center };
-    { _At._On_width(_STD declval<int>()) } -> same_as<void>;
-    { _At._On_dynamic_width(_STD declval<int>()) } -> same_as<void>;
-    { _At._On_dynamic_width(_STD declval<_Auto_id_tag>()) } -> same_as<void>;
-    { _At._On_precision(_STD declval<int>()) } -> same_as<void>;
-    { _At._On_dynamic_precision(_STD declval<int>()) } -> same_as<void>;
-    { _At._On_dynamic_precision(_STD declval<_Auto_id_tag>()) } -> same_as<void>;
+    { _At._On_width(static_cast<unsigned int>(0)) } -> same_as<void>;
+    { _At._On_dynamic_width(size_t{}) } -> same_as<void>;
+    { _At._On_dynamic_width(_Auto_id_tag{}) } -> same_as<void>;
+    { _At._On_precision(static_cast<unsigned int>(0)) } -> same_as<void>;
+    { _At._On_dynamic_precision(size_t{}) } -> same_as<void>;
+    { _At._On_dynamic_precision(_Auto_id_tag{}) } -> same_as<void>;
     { _At._On_sign(_Sgn) } -> same_as<void>;
     { _At._On_hash() } -> same_as<void>;
     { _At._On_zero() } -> same_as<void>;
-    { _At._On_type(_STD declval<_CharT>()) } -> same_as<void>;
+    { _At._On_type(_CharT{}) } -> same_as<void>;
 };
 template <class _Ty, class _CharT>
 concept _Parse_arg_id_callbacks = requires(_Ty _At) {
     { _At._On_auto_id() } -> same_as<void>;
-    { _At._On_manual_id(_STD declval<int>()) } -> same_as<void>;
+    { _At._On_manual_id(size_t{}) } -> same_as<void>;
 };
+
+template<class _Ty, class _CharT>
+concept _Parse_replacement_field_callbacks = requires(_Ty _At, const _CharT* _Begin, const _CharT* _End) {
+    { _At._Parse_context };
+    { _At._On_text(_Begin, _End) } -> same_as<void>;
+    { _At._On_replacement_field(size_t{}, static_cast<const _CharT*>(nullptr)) } -> same_as<void>;
+    { _At._On_format_specs(size_t{}, _Begin, _End) } -> same_as<const _CharT*>;
+};
+
 // clang-format on
 
+template <class _Ty, class _CharT = char>
+struct formatter;
+
+inline void _You_see_this_error_because_arg_id_is_out_of_range() noexcept {}
+
+template <class _CharT>
+class basic_format_parse_context {
+public:
+    using char_type      = _CharT;
+    using const_iterator = typename basic_string_view<_CharT>::const_iterator;
+    using iterator       = const_iterator;
+
+    constexpr explicit basic_format_parse_context(basic_string_view<_CharT> _Fmt, size_t _Num_args_ = 0) noexcept
+        : _Format_string(_Fmt), _Num_args(_Num_args_) {}
+    basic_format_parse_context(const basic_format_parse_context&) = delete;
+    basic_format_parse_context& operator=(const basic_format_parse_context&) = delete;
+
+    _NODISCARD constexpr const_iterator begin() const noexcept {
+        return _Format_string.begin();
+    }
+    _NODISCARD constexpr const_iterator end() const noexcept {
+        return _Format_string.end();
+    }
+    constexpr void advance_to(const const_iterator _It) {
+        _Adl_verify_range(_It, _Format_string.end());
+        const auto _Diff = static_cast<size_t>(_It._Unwrapped() - _Format_string._Unchecked_begin());
+        _Format_string.remove_prefix(_Diff);

```

Comment:
I'm concerned that `_Diff` could be negative here and `remove_prefix` will do the wrong thing (it takes a `size_t`). We could we comment this for now and worry about it later.